### PR TITLE
Implement new behavior for OnDriverDistraction 

### LINKED ISSUE #1881: [OnDriverDistraction] SDL does not send notification to app right after this app changes level from NONE to any other
## Occurrence: 
Always 

## Steps to reproduce:
1. Make sure SDL is built with default flags
2. Start SDL and HMI
3. Connect mobile device
4. Register new application → it gets NONE HMI level and SDL receives OnDriverDistraction (<state>) fom HMI 
5. Change HMI level to any other than NONE: (FULL, BACKGROUND, LIMITED)
6. Check that 'OnDriverDistraction' notification is sent to application

## Expected result:
SDL must transfer the last known (actual) OnDriverDistraction (<state>) to the mobile app right after this mobile app changes HMI level to any other than NONE.

## Actual result:
N/A

## Test script:
[1881_OnDriverDistraction_After_changing_HMIlevel_from_NONE.lua](https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/develop/test_scripts/Defects/4_5/1881_OnDriverDistraction_After_changing_HMIlevel_from_NONE.lua)

## Attachments:
Note: Extract attached folder with script to test_scripts folder. 
[Defects.zip](https://github.com/smartdevicelink/sdl_core/files/1517516/Defects.zip)

# Expected delivery
- [ ] Source code updates
- [ ] Code comments
- [x] **UTs add/update _(not required)_**
- [ ] ATF tests add/update
- [x] **Manual tests _(not required)_**
- [x] **Add/update CI plans/jobs _(not required)_**
- [x] **SDD updates _(not required)_**
- [x] **Guidelines update ([sdl_core_guides](https://github.com/smartdevicelink/sdl_core_guides)) _(not required)_**
- [x] **Guidelines update ([sdl_hmi_integration_guidelines](https://github.com/smartdevicelink/sdl_hmi_integration_guidelines)) _(not required)_**

### PR DESCRIPTION
**Pull request by LitvinenkoIra**
Fixes #1881

This PR is **[ready]** for review.

### Risk
This PR makes **[no]** API changes.

### Testing Plan
Updated appropriate unit tests

### Summary
According to the new behavior SDL has to postpone OnDriverDistraction
for the application which are in NONE HMI level and transfer the last known (actual) OnDriverDistraction (<state>) to this mobile app right after this mobile app changes HMILevel to any other than NONE.
Note: Whether the app receives the notification in current HMILevel is defined by app's assigned Policies. Often, OnDriverDistraction is disallowed for apps in NONE. But if OnDriverDistraction is allowed for HMI status = NONE, SDL must transfer this notification to appication.

### Changelog
- Added function that puts OnDriverDistraction notification to postponed 
- Added function that processes this notification when app changes HMILevel to any other than NONE.
 
### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

(first review iteration: https://github.com/LitvinenkoIra/sdl_core/pull/2)

### CODE REVIEW

File: src/components/application_manager/src/application_manager_impl.cc [hunk 1]
```diff
@@ -3889,6 +3927,28 @@ void ApplicationManagerImpl::OnPTUFinished(const bool ptu_result) {
 #endif  // SDL_REMOTE_CONTROL
 }
 
+void ApplicationManagerImpl::PutDriverDistractionMessageToPostponed(
+    ApplicationSharedPtr application) const {
+  LOG4CXX_AUTO_TRACE(logger_);
+  if (hmi_apis::Common_DriverDistractionState::INVALID_ENUM ==
+      distracting_driver_state()) {
+    LOG4CXX_INFO(logger_, "DriverDistractionState is INVALID_ENUM");
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/application_manager_impl.cc]: @LitvinenkoIra please do not use INFO level loggin for such logs. Consider https://github.com/smartdevicelink/sdl_core/wiki/SDL-Logging-levels

Comment (LitvinenkoIra) [reply src/components/application_manager/src/application_manager_impl.cc]: @LuxoftAKutsan done 212bee04d27efd4be438aa8b57966dee010fda8e

File: src/components/application_manager/include/application_manager/application.h [hunk 2]
```diff
@@ -168,6 +168,11 @@ typedef std::set<mobile_apis::VehicleDataType::eType> VehicleInfoSubscriptions;
  */
 typedef std::set<mobile_apis::ButtonName::eType> ButtonSubscriptions;
 
+/**
+ * @breif Colelction for the mobile command smart object.
```
Comment (AByzhynar) [inline src/components/application_manager/include/application_manager/application.h]: @LitvinenkoIra typo

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 3]
```diff
@@ -463,14 +463,16 @@ class ApplicationManagerImpl
    *
    * @return Current state of the distraction state
    */
-  inline bool driver_distraction() const;
+  inline hmi_apis::Common_DriverDistractionState::eType
```
Comment (AByzhynar) [inline src/components/application_manager/include/application_manager/application_manager_impl.h]: @LitvinenkoIra We don't use `inline`'

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 4]
```diff
@@ -1353,6 +1355,32 @@ class ApplicationManagerImpl
    */
   bool IsLowVoltage();
 
+  /**
+   * @brief Allows to process postponed commands for application
+   * when itshmi level has been changed.
```
Comment (AByzhynar) [inline src/components/application_manager/include/application_manager/application_manager_impl.h]: @LitvinenkoIra typo

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 5]
```diff
@@ -1445,6 +1473,14 @@ class ApplicationManagerImpl
    */
   protocol_handler::MajorProtocolVersion SupportedSDLVersion() const;
 
+  /**
+   * @brief Checks if driver distraction state is valid, create message
```
Comment (AByzhynar) [inline src/components/application_manager/include/application_manager/application_manager_impl.h]: @LitvinenkoIra creates

File: src/components/application_manager/src/application_manager_impl.cc [hunk 6]
```diff
@@ -3405,6 +3408,41 @@ void ApplicationManagerImpl::OnHMILevelChanged(
...
+  app->SwapMobileMessageQueue(messages);
+  auto push_allowed_messages =
+      [this, &app](smart_objects::SmartObjectSPtr message) {
+        const std::string function_id = MessageHelper::StringifiedFunctionID(
+            static_cast<mobile_apis::FunctionID::eType>(
+                (*message)[strings::params][strings::function_id].asUInt()));
+        const RPCParams params;
+        const mobile_apis::Result::eType check_result =
+            CheckPolicyPermissions(app, function_id, params);
+        if (check_result == mobile_api::Result::SUCCESS) {
```
Comment (AByzhynar) [inline src/components/application_manager/src/application_manager_impl.cc]: @LitvinenkoIra I would recommend to swap variables in comparison

File: src/components/application_manager/src/application_manager_impl.cc [hunk 7]
```diff
@@ -3405,6 +3408,41 @@ void ApplicationManagerImpl::OnHMILevelChanged(
...
+          ManageMobileCommand(message,
+                              commands::Command::CommandOrigin::ORIGIN_SDL);
+        } else {
+          app->PushMobileMessage(message);
+        }
+      };
```
Comment (AByzhynar) [inline src/components/application_manager/src/application_manager_impl.cc]: @LitvinenkoIra Do we need `;` here?

File: src/components/application_manager/src/commands/hmi/on_driver_distraction_notification.cc [hunk 8]
```diff
@@ -34,15 +34,64 @@
 #include "application_manager/commands/hmi/on_driver_distraction_notification.h"
 
 #include "application_manager/application_impl.h"
+#include "application_manager/message_helper.h"
 #include "interfaces/MOBILE_API.h"
 #include "interfaces/HMI_API.h"
+#include "utils/make_shared.h"
```
Comment (AByzhynar) [inline src/components/application_manager/src/commands/hmi/on_driver_distraction_notification.cc]: @LitvinenkoIra What for?

File: src/components/application_manager/src/commands/hmi/on_driver_distraction_notification.cc [hunk 9]
```diff
@@ -34,15 +34,64 @@
...
+      }
+      command_.SendNotificationToMobile(on_driver_distraction_so_);
+    }
+  }
+
+ private:
+  OnDriverDistractionNotification& command_;
+  smart_objects::SmartObjectSPtr on_driver_distraction_so_;
+  ApplicationManager& application_manager_;
+  std::string stringified_functionID_;
```
Comment (AByzhynar) [inline src/components/application_manager/src/commands/hmi/on_driver_distraction_notification.cc]: @LitvinenkoIra 
- named wrongly according Google Coding Style

File: src/components/application_manager/src/commands/hmi/on_driver_distraction_notification.cc [hunk 10]
```diff
@@ -51,38 +100,31 @@ OnDriverDistractionNotification::~OnDriverDistractionNotification() {}
...
-
   const hmi_apis::Common_DriverDistractionState::eType state =
       static_cast<hmi_apis::Common_DriverDistractionState::eType>(
           (*message_)[strings::msg_params][hmi_notification::state].asInt());
-  application_manager_.set_driver_distraction(state);
+  application_manager_.set_distracting_driver_state(state);
 
   smart_objects::SmartObjectSPtr on_driver_distraction =
-      new smart_objects::SmartObject();
+      utils::MakeShared<smart_objects::SmartObject>();
```
Comment (AByzhynar) [inline src/components/application_manager/src/commands/hmi/on_driver_distraction_notification.cc]: @LitvinenkoIra std::make_shared?

File: src/components/application_manager/test/application_impl_test.cc [hunk 11]
```diff
@@ -784,6 +784,29 @@ TEST_F(ApplicationImplTest, StopStreaming_StreamingApproved) {
   EXPECT_FALSE(app_impl->audio_streaming_approved());
 }
 
+TEST_F(ApplicationImplTest, PushPopMobileMessage) {
+  smart_objects::SmartObjectSPtr on_driver_distraction =
+      utils::MakeShared<smart_objects::SmartObject>();
```
Comment (AByzhynar) [inline src/components/application_manager/test/application_impl_test.cc]: @LitvinenkoIra std::make_shared?

File: src/components/application_manager/src/application_manager_impl.cc (hunk 7 above)
Comment (LitvinenkoIra) [reply src/components/application_manager/src/application_manager_impl.cc]: @AByzhynar yes

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 12]
```diff
@@ -463,14 +463,16 @@ class ApplicationManagerImpl
    *
    * @return Current state of the distraction state
    */
-  inline bool driver_distraction() const;
+  inline hmi_apis::Common_DriverDistractionState::eType
+  distracting_driver_state() const;
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/include/application_manager/application_manager_impl.h]: @LitvinenkoIra propose to rename to `driver_distraction_state()` / `set_driver_distraction_state()`

File: src/components/application_manager/src/application_manager_impl.cc (hunk 6 above)
Comment (LitvinenkoIra) [reply src/components/application_manager/src/application_manager_impl.cc]: @AByzhynar Done e638bda5ff050f5c1d32aeb2dc575c1016170ca1

File: src/components/application_manager/include/application_manager/application_manager_impl.h (hunk 5 above)
Comment (LitvinenkoIra) [reply src/components/application_manager/include/application_manager/application_manager_impl.h]: @AByzhynar done e638bda5ff050f5c1d32aeb2dc575c1016170ca1

File: src/components/application_manager/include/application_manager/application_manager_impl.h (hunk 4 above)
Comment (LitvinenkoIra) [reply src/components/application_manager/include/application_manager/application_manager_impl.h]: @AByzhynar fixed e638bda5ff050f5c1d32aeb2dc575c1016170ca1

File: src/components/application_manager/include/application_manager/application.h (hunk 2 above)
Comment (LitvinenkoIra) [reply src/components/application_manager/include/application_manager/application.h]: @AByzhynar fixed e638bda5ff050f5c1d32aeb2dc575c1016170ca1

File: src/components/application_manager/src/commands/hmi/on_driver_distraction_notification.cc (hunk 9 above)
Comment (LitvinenkoIra) [reply src/components/application_manager/src/commands/hmi/on_driver_distraction_notification.cc]: @AByzhynar fixed e638bda5ff050f5c1d32aeb2dc575c1016170ca1

File: src/components/application_manager/include/application_manager/application_manager_impl.h (hunk 12 above)
Comment (LitvinenkoIra) [reply src/components/application_manager/include/application_manager/application_manager_impl.h]: @AKalinich-Luxoft Done e638bda5ff050f5c1d32aeb2dc575c1016170ca1

**Pull request merged**